docs/feature/dualstack/Cilium-dualstack.md

```diff
@@ -0,0 +1,67 @@
+1. Azure-ipam needs to be able to assign multiple IPs
+2. CNS needs to be able to determine what type of IP it is returning so that we can have one IPv4 and one IPv6. 
+
+## Current Implementation
+
+When we use Cilium CNI with Azure CNS we use an Azure-ipam plugin to communicate from the Cilium CNI and CNS. This plugin takes in add commands from the Cilium CNI and reads the CNI network config from stdin. The data from stdin then gets parsed into an [`IPConfigRequest`](https://github.com/Azure/azure-container-networking/blob/master/cns/NetworkContainerContract.go) object and is passed to [`RequestIPAddress`](https://github.com/Azure/azure-container-networking/blob/master/cns/client/client.go). From here we call the [`requestIPConfigHandler`](https://github.com/Azure/azure-container-networking/blob/master/cns/restserver/ipam.go#:~:text=func%20(service%20*HTTPRestService)%20requestIPConfigHandler(w%20http.ResponseWriter%2C%20r%20*http.Request)%20%7B) in ipam using which then retrieves and validates the `IPConfigRequest` and then calls the [`requestIPConfigHelper`](https://github.com/Azure/azure-container-networking/blob/master/cns/restserver/ipam.go#:~:text=func%20requestIPConfigHelper(service%20*HTTPRestService%2C%20req%20cns.IPConfigRequest)%20(cns.PodIpInfo%2C%20error)%20%7B). The helper function first checks to see if the `IPConfigRequest` already belongs to a pod and returns the existing pod's information. If it doesn't already exist than it checks if there is a desiredIPaddress or not. If there is a desired IP address than we search for that specific IP and if not it calls 
+[`AssignAnyAvailableIPConfig`](https://github.com/Azure/azure-container-networking/blob/master/cns/restserver/ipam.go#:~:text=func%20(service%20*HTTPRestService)%20AssignAnyAvailableIPConfig(podInfo%20cns.PodInfo)%20(cns.PodIpInfo%2C%20error)%20%7B)
+## New Implementation
+
+### NNC change
+
+For the CNS to know that we are in dual stack we will use a flag that will be passed in using the NNC. This will be filled in from the DNC-RC side and then sent to the CNS over the CRD in a flag called `IPMode`. The IPMode flag will be of type IPMode and will have two options "SingleStack" which when passed in will have the CNS work as it currently does and "Dual Stack" which tell the CNS to look for two IPs instead of one.
+
+```diff
+// NodeNetworkConfigStatus defines the observed state of NetworkConfig
+type NodeNetworkConfigStatus struct {
+	AssignedIPCount   int                `json:"assignedIPCount,omitempty"`
+	Scaler            Scaler             `json:"scaler,omitempty"`
+	Status            Status             `json:"status,omitempty"`
++   IPMode            IPMode             `json:"IPmode,omitempty"`
+	NetworkContainers []NetworkContainer `json:"networkContainers,omitempty"`
+}
+```
+
+```diff
++type IPMode string
++
++const (
++	SingleStack Status = "SingleStack"
++	DualStack   Status = "DualStack"
++)
```

Comment:
I don't think that this is the concept that we previously discussed and agreed on - I want the NC assignment mode to be more generic and composable. You are trying to address the case of "one v4 NC, one v6 NC, and assign an IP from both", but we could more fundamentally describe that as "two NCs, assign an IP from both". Then, what if we have more than two NCs? This is not different from "multiple NCs, assign an IP from each". What about the scenario of all NCs are v4, or all are v6? What if I have several v4 and one v6?
In other words, you are addressing the _specific case_ of v4+v6, but you should be thinking about how to solve the general problem of CNS having multiple NCs and needing to assign IPs out of one or more of them in certain combinations (and hardcoding the behavior of each option is the unmaintainable way to do it). The solution you propose can still be to the specific case you are interested in, but should be open-ended enough that it could be enhanced to the more general cases without having to throw it away and re-solve the whole problem again.

To get this started, we should initially change the option name to `IPAssignmentMode` or `NCAssignmentMode`, and think about how we tell CNS _dynamically, at runtime_ which NC(s) it should be assigning an IP out of to satisfy any individual IP request.

Reply:
Hey @rbtr, the last conversation us three had, we talked about having a single field similar to what Ryan proposed here to let CNS know that it needs to operate in the dual stack mode. Maybe there was a misunderstanding on takeaways from that discussion. Having said that, I see your point that about adding a field that couples closely to a scenario. I don't believe we want to design something for the scenarios which are not existent and don't know what the usage will be. e.g. having multiple NCs in NNCs - can there be mixed mode supported where few NCs will be dynamic, few static? In this case, CNS would not want to give IPs from both the NCs as suggested by NNC's IPAssignmentMode NC list. Will there be a valid scenario where there would be say 5 NCs under NNC while CNS is supposed to use only 3 of those to generate IP pool? etc. There are unknowns and IMO we should design for those when we actually have valid scenarios.
In short, I am proposing to stick to the current behavior of CNS to use all the NCs under NNC to populate the IP pool and not add any flag to the NNC.
Proposal -
1. CNS learns IPAssignmentMode based on IPFamily and mode (static/dynamic) from the NCs under NNC. If CNS sees ipv4 and ipv6 static mode NCs under NNC, it will be in dualstack IPAssignmentMode. In this mode, it will validate that there are exactly 2 NCs under NNC.
2. When in dualstack mode, CNS will give out one ipv4 and one ipv6 IP.
3. There won't be any change to ipv4 overlay for now but this can be extended to it as well.

Let me know your thoughts.

Reply:
If we want to go that direction, I suggest we commit to it fully: we can remove any intelligence related to IPv4/v6 and make it so that CNS assigns one IP out of each NC for each IP Request. 
This is the simplest solution to the current problem and requires the least rework in the future if other NC scenarios come up, because we won't have to _remove_ a bunch of logic around how we assign IPs based on the IP families of the NCs.

Reply:
We don't know what future NC scenarios will come up. If there is a scenario of mix mode, then we still may not want to assign one IP out of each NC for each IP request. So if we go with this route we may still need to remove / update logic. We just don't know at this point :) 
I am ok with giving one IP out of each NC and removing any intelligence about ip familiy. Without digging into the CNS code, do you know if that will affect the current logic of v4 overlay? I'll try to take a look at it

Reply:
No existing singletenancy scenario has multiple NCs so they should be fine